### ISSUE TITLE
Sets should be re written using Red Black Trees. 

### ISSUE THREAD

**Issue by professorcode1** [igraph/igraph#2287]
Sets are currently implemented as an array of integers. When a new element is inserted binary search is used to find its position and the entire array is shifted. This a sub-optimal performance vice.  Also the state is just an index in the array, this means if an element is added in the set while use are iterating before it then the element will be read twice. This is a bug, to replicate it
```
    igraph_set_t set;
    IGRAPH_CHECK(igraph_set_init(&set, 0));
    IGRAPH_FINALLY(igraph_set_destroy, &set);

    igraph_set_add(&set,1);
    igraph_set_add(&set,2);
    igraph_set_add(&set,3);
    igraph_set_add(&set,4);
    igraph_set_add(&set,5);

    igraph_integer_t state, element;
    for(state=0;igraph_set_iterate(&set, &state,&element) ;){
        if(element==3){
            igraph_set_add(&set, 0);
        }
        printf("%ld\n", element);
    }
``` 
3 will get printed twice. 
There is also no functionality to remove elements from the set.  And if there was one added now it would also become a bug because of how the iterate function works. Hence the set class needs to be re written properly. 

To allow for all POD I intend to use the methods and API described here [here](https://github.com/igraph/igraph/issues/2285). 

**Use Case**
Complex graph algorithms like ant-colony coloring need sets to be implemented, and are a good tool to have in general. To keep the current code that use the sets working, a wrapper can be make around the RBT to make a set that only store integers.

**References**
CLSR has a chapter on how to implement red and black tress. 

Please assign the issue to me.

**Comment by ntamas** [comment]
I was wondering whether we should simply make use of many of the existing generic header-only C implementations of sets instead of rolling our own. E.g., here is this one:

https://github.com/rurban/ctl/blob/master/docs/set.md

**Comment by professorcode1** [comment]
Please let me know once it is merged into master on the ant colony thread.

**Comment by ntamas** [comment]
oh sorry, I meant to assign this to you; it looks like I was quite tired last night when I did this. Feel free to decide how to go ahead with the set implementation, I was only suggesting that maybe it's not worth reinventing the wheel if there are suitable solutions out there.

**Comment by szhorvat** [comment]
> To allow for all POD I intend to use the methods and API described [here](https://github.com/igraph/igraph/issues/2285).
> EDIT: by that I mean I intend to create an API similar to this one and use the same void pointer method to emulate comparison overloading.

This is of course useful in that it allows storing arbitrary objects in sets. However, it is likely to be significantly slower than an implementation specialized for storing integers, as the comparison operation cannot be inlined. Therefore, realistically, we'll probably continue to have at least a separate integer-only implementation.

**Comment by professorcode1** [comment]
I agree @ntamas that its a bad idea to reinvent the wheel. In order to use it do I need to  port it (like rewrite it but replace arrays with igraph vectors and malloc with igraph_malloc and such) or is it okay to just include that file as is in the source code?
@szhorvat I won't remove the current version of set that is in the code then .

**Comment by ntamas** [comment]
I would probably create an `igraph_set_t` that has an identical interface to what we have now but uses a well-established generic solution under the hood, preferably optimized for integers.

**Comment by professorcode1** [comment]
I just noticed, there already is an implementation of sets for integers in src/cliques/cliquer/set.h. And the way its implemented it looks to me like it doesn't suffer from the same bugs that the current set implementation does. It also seems optimised for ints. How about I just copy this into the current core/set.c make some changes so it fits nicely with the API and call it a day?

**Comment by ntamas** [comment]
`src/cliques/cliquer/set.h` is a set implemented with bitfields, i.e. if you want to store integers between 0 and `n-1`, you'll need `ceil(n/8)` bytes for that. This fits a very specific use-case in the Cliquer algorithm, but I don't think it's a good fit for a "generic" set. _If_ it fits your particular use-case for the ant colony optimization algorithm, I guess you can borrow the implementation by including its header directly, but I wouldn't make it a replacement of `igraph_set_t`; I'm pretty sure the next feature request from someone else would be something like "`igraph_set_t` uses too much memory, it should be re-written using red-black trees" :)

**Comment by professorcode1** [comment]
@ntamas  The ctl library requires that there be a compiler variable T defined to store the data type, so to use it one would have to define T as a compiler directive, do what they wanna do then undefine it which seems messy(I don't even know if compiler variables can be undefined). [Here](https://github.com/20centaurifux/datatypes) is an implementation that I found. Its great. Can I port this?

**Comment by szhorvat** [comment]
As  I said [here](https://github.com/igraph/igraph/pull/2284#issuecomment-1368208735), please let's do things in order. We are not yet at the point of having to do this.

**Comment by professorcode1** [comment]
So should I port this now?

**Comment by ntamas** [comment]
I've looked into the current `igraph_set_t` code a bit to jog my memory. It seems like `igraph_set_t` is currently a "private export", i.e. it is exported from the library but it is not documented and not meant to be used by end users. So we can be a bit less constrained in how we change the internals. My preference would be to check whether the set data structure can be improved, but with the following constraints:

* Let's keep the existing API if possible and only re-write the internals.
* Add a benchmark that exercises some part of the igraph code that uses sets heavily to gauge whether we gain any improvement in a real use-case by changing the internals of the set. Candidates would be `igraph_induced_subgraph_edges()` or some specific methods of `igraph_degree_sequence_game()`. Changing the internals of `igraph_set_t` without measuring whether we gain something is probably too much work.

So, start with a benchmark first. You can find the benchmarks in `tests/benchmarks`; there is a file named `bench.h` with a bunch of utility functions that lets you concentrate on writing the code to be measured and takes care of the actual time measurements and formatting of the output.

**Comment by professorcode1** [comment]
I'll do the benchmarks, but the code is bugged so it has to be changed. I only need integer sets,  so I'll only add those (lets forget about the all possible pod's plan) without changing the api. Though it would be challenge since current iterator state is an int. One way would be to cast a pointer of a tree node to igraph_integer_t since they are ints. Another way would be make a igraph_set_iterator_t and changed the only two files that use set iterator (cliques.c and degree_sequence.c). Which would you prefer?   I prefer the later.

**Comment by ntamas** [comment]
> the code is bugged so it has to be changed.

No problem, if there are bugs, we'd better fix them anyway.

> I only need integer sets, so I'll only add those 

:+1:

> Though it would be challenge since current iterator state is an int.

Since `igraph_set_t` is not part of the public API, we can bend the rules here a bit and change the interface. Instead of taking an `igraph_integer_t*` as the state, `igraph_set_iterate` should take an `igraph_set_iterator_t` as you suggested. The API should be modelled similarly to existing iterator data types like `igraph_sparsemat_iterator_t` or `igraph_2dgrid_iterator_t`. In particular, I'd recommend:

```
/* Resets an iterator to the start of the given set */
igraph_error_t igraph_set_iterator_reset(const igraph_set_t* set, igraph_set_iterator_t* it);

/* Advances the iterator to the next element; returns true if the iterator was not at the end before calling the function, false if it was already at the end of the set */
igraph_bool_t igraph_set_iterator_next(igraph_set_iterator_t* it);

/* Returns whether the iterator is already at the end, i.e. past the last element */
igraph_bool_t igraph_set_iterator_end(const igraph_set_iterator_t* it);

/* Returns the current element of the iterator; behaviour undefined if the iterator is already at the end (i.e. past the last element) */
igraph_integer_t igraph_set_iterator_get(const igraph_set_iterator_t* it);
```

This way one could iterate over a set with a `while` loop:

```
IGRAPH_CHECK(igraph_set_iterator_reset(&set, &it));
while (!igraph_set_iterator_end(&it)) {
    igraph_integer_t current = igraph_set_iterator_get(&it);
    igraph_set_iterator_next(&it);   /* guaranteed to succeed */
}
```

@szhorvat Any thoughts about the set iterator API? In particular:

* Should the `_next()` method attempt to return an `igraph_error_t` instead?
* Should the `_get()` method return an `igraph_error_t` and provide an output argument instead?
* Note the ordering of the arguments in `_reset()`; this is to remain consistent with the already existing `_reset()` methods.

I'm trying to come up with an API that is convenient for `while` and `for` loops. That's why I decided to return an `igraph_bool_t` in most cases, otherwise we would be forced to use an `IGRAPH_CHECK()` around these functions, which does not play nicely with `for` loops.

**Comment by professorcode1** [comment]
@ntamas I am implementing sets using [this](https://github.com/amitbansal7/Data-Structures-and-Algorithms/blob/master/9.Red-Black-tree/RedBlackTrees.c) as reference and making the necessary changes. Please let me know now about the API changes now if you would like any.

**Comment by professorcode1** [comment]
Hey! Someone please tell me how to handle the reserve size passed to the set in the new implementation.
Right now I am keeping a reservoir of Nodes and I call realloc when the size exceeds the capacity. You can check it [here](https://github.com/igraph/igraph/pull/2313) . The problem is the realloc makes the tree invalid since now the nodes point at locations that are not correct.
I can think of a few solutions , below are the order in which I like them
1) Make the nodes stores the index of the nodes in the reservoir instead of their memory address.
2) Make the reservoir an array of arrays of Nodes, and every time the capacity is exceeded make a new array of nodes. 
3) Update the entire tree inplace on each realloc
4) Drop the reserve functionality 
Please let me know how to proceed @ntamas  @szhorvat

**Comment by professorcode1** [comment]
Requesting an update on this!

**Comment by ntamas** [comment]
You can drop the `_reserve()` function if that complicates things; I think it is not needed in this case. Typically, `reserve()`-style functions are used in igraph's API when the data structure is backed by an internal array and you want to give a hint to the data structure about the number of elements it should expect in the near future. It's not a big deal if we don't have it here.

**Comment by professorcode1** [comment]
The sets pull request is ready for its first review.

**Comment by ntamas** [comment]
PR #2313 is still ongoing but it's unclear yet what the outcome will be. Nevertheless, I'm removing the stale label.
